File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/usecase/ListValorantSkinsByThemeUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.usecase;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.stereotype.Service;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.ports.ValorantSkinRepository;
import fr.huiitre.tools.modules.riot.valorant.application.view.ValorantSkinView;

@Service
public class ListValorantSkinsByThemeUseCase implements SecuredUseCase {

    private final ValorantSkinRepository skinRepository;

    public ListValorantSkinsByThemeUseCase(ValorantSkinRepository skinRepository) {
        this.skinRepository = skinRepository;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.READ_ONLY;
    }

    public List<ValorantSkinView> execute(UUID themeUuid) {
        return skinRepository.findAllByTierUuid(themeUuid);
    }
}

File: api/Modules/Riot/Valorant/Application/Catalog/Usecases/ListValorantSkinsByThemeUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Skin.Views;

namespace Tools.Api.Modules.Riot.Valorant.Application.Catalog.Usecases;

public sealed class ListValorantSkinsByThemeUseCase(
    UseCaseAuthorizer authorizer,
    IValorantAuthRepository valorantAuthRepository,
    IValorantSkinRepository skinRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.ReadOnly;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    public async Task<List<ValorantSkinView>> Execute(Guid themeUuid, long? accountId)
    {
        // accountId est facultatif, mais s'il est fourni il doit appartenir à l'appelant : sinon
        // « possédé » et « suivi » seraient lus sur le compte d'un autre.
        if (accountId is { } linkedAccountId && !await valorantAuthRepository.ExistsByIdAndUserId(linkedAccountId, CurrentUser.UserId))
        {
            throw AppException.NotFound(
                "VALORANT_ACCOUNT_NOT_FOUND",
                "Ce compte Valorant est introuvable.");
        }

        return await skinRepository.FindAllByTierUuid(themeUuid, accountId);
    }
}
